### DISCUSSION TITLE
GGUF huge files size / Split possible ?

### DISCUSSION THREAD

**Discussion by phymbert** [ggml-org/llama.cpp#6000] [Q&A] [answered]
Distributing and storing GGUF files is difficult for 70b models, especially on f16. Lot of issue can happen during file transfers, examples:
 - temporary disk full
 - network interruption
 
Typically,  they need to be tranferred from huggingface to an internal storage like s3, minio, git lfs, nexus or artifactory, then downloaded by the inference server and stored locally (or on a k8s PvC for example). Also they cannot be stored in a dockerfile, but IMHO this is for good.

Is there a smarter way than splitting the file using `split` & `cat` ?

**Comment by phymbert** [comment]
@ggerganov thoughts? would it be possible to support a split size in GUFF ?

**Reply by ggerganov** [reply]
We can give it a try - I can't estimate how big would be the change. A rough plan would be:

- implement `split` tool that given a GGUF file generate `N` GGUF files
- modify `llama_model_loader` to load tensors from multiple GGUf files (probably checking for extension `.gguf.1`, `.gguf.2`, etc)
- maybe add optional meta data in the GGUF to indicate number of splits

Ideally, no changes to `ggml` would be needed.

Depending on the complexity of the implementation we will decide wether to support or not

**Reply by phymbert** [reply]
Ok, I will give it a try. Thanks for the guidance

**Reply by ngxson** [reply]
Good idea!

I've just had a look on `llama_model_loader`. What I understand is that we can add one tuple `gguf_context, ggml_context` for each file. We can then have a `std::unordered_map<std::string, ggml_context>` that maps the tensor name to the correct context.

One thing that I'm not sure is that `create_tensor` and `create_tensor_for` are currently taking both `ggml_context` and tensor name as input - this maybe tricky because now we need to either select the context outside of `create_tensor`, or either just pass the mapping to `create_tensor` and let it search for the context.

Maybe there's better idea than mine?

**Reply by slaren** [reply]
The `ggml_context` that `create_tensor` and `create_tensor_for` takes is just an user-provided context where the tensor should be created, you don't need to do anything special with these. Just find the `ctx_meta` where the tensor is, and duplicate it from there into the user-provided context.

**Reply by phymbert** [reply]
Getting the tensor is OK, loading the tensor data from multiple files is more complicated especially with `mmap`. Any help is welcomed based on 14442d8315f5522d8d71cc20f55bdbda6f35bc72

**Comment by Artefact2** [comment]
> Is there a smarter way than splitting the file using split & cat ?

On Linux yes, if your filesystem supports reflink copies (btrfs, xfs do). I've written a couple demo utilities that use `copy_file_range()` for near-instant merging/splitting of files regardless of their size.

<details>
<summary>my-split.c</summary>

```c
#define _GNU_SOURCE
#define _FILE_OFFSET_BITS 64

#include <stdio.h>
#include <fcntl.h>
#include <unistd.h>
#include <string.h>
#include <unistd.h>

int main(int argc, char** argv) {
    if(argc != 2) {
        fprintf(stderr, "Usage: %s file.gguf\n", argv[0]);
        return 1;
    }

    int src_fd = open(argv[1], O_RDONLY);
    if(src_fd == -1) {
        perror("src open()");
        return 1;
    }
    off_t src_len = lseek(src_fd, 0, SEEK_END);
    if(src_len == -1) {
        perror("lseek()");
        return 1;
    }
    lseek(src_fd, 0, SEEK_SET);

    /* 50GB (NOT GiB) */
    const off_t BSIZE = 372*128*1024*1024l; /* Must be a multiple of CFRSIZE */
    const size_t CFRSIZE = 128*1024*1024l;
    if(src_len < BSIZE) return 0; /* No need to split */

    unsigned int n = 0;
    char dest_path[strlen(argv[1]) + 10];
    int dest_fd;
    for(off_t i = 0; i < src_len; i += BSIZE) {
        snprintf(dest_path, sizeof(dest_path)-1, "%s.%03u", argv[1], n);
        dest_fd = open(dest_path, O_CREAT|O_TRUNC|O_WRONLY, 0666);
        if(dest_fd == -1) {
            perror("dest open()");
            return 1;
        }
        for(off_t k = 0; k < BSIZE; k += CFRSIZE) {
            if(copy_file_range(src_fd, NULL, dest_fd, NULL, CFRSIZE, 0) == -1) {
                perror("copy_file_range()");
                return 1;
            }
        }
        close(dest_fd);
        printf("%s\n", dest_path);
        n += 1;
    }
    close(src_fd);
    return 0;
}
```

</details>

<details>
<summary>my-merge.c</summary>

```c
#define _GNU_SOURCE
#define _FILE_OFFSET_BITS 64

#include <stdio.h>
#include <fcntl.h>
#include <unistd.h>
#include <string.h>
#include <unistd.h>

int main(int argc, char** argv) {
    if(argc < 3) {
        fprintf(stderr, "Usage: %s file.000 file.001... merge.out\n", argv[0]);
        return 1;
    }

    int dest_fd = open(argv[argc-1], O_CREAT|O_TRUNC|O_WRONLY, 0666);
    if(dest_fd == -1) {
        perror("dest open()");
        return 1;
    }
    const size_t CFRSIZE = 2048*1024*1024l;

    for(int i = 1; i < argc-1; ++i) {
        int in_fd = open(argv[i], O_RDONLY);
        if(in_fd == -1) {
            perror("in open()");
            return 1;
        }

        ssize_t ret;
        /* copy_file_range() returns 0 on EOF */
        while((ret = copy_file_range(in_fd, NULL, dest_fd, NULL, CFRSIZE, 0)) != 0) {
            if(ret == -1) {
                perror("copy_file_range()");
                return 1;
            }
        }
        close(in_fd);
    }

    close(dest_fd);
    return 0;
}
```
</details>

**Reply by ngxson** [reply]
What I'm not sure (in theory) is that will there be any performance cost if the file is splitted in the middle of a tensor?

That means for example, a file having `tensor0 tensor1 tensor2`

Now it's splitted into 2 files:
- `tensor0 tensor1(first-half)`
- `tensor1(second-half) tensor2`

With `copy_file_range` it will not consolidate these "fragments" of a file, but just link them together. Maybe when `mmap` tries to load `tensor1` to memory, it will need to make a copy of `tensor1` into RAM?

**Reply by Artefact2** [reply]
No, that's not how it works. My code is semantically equivalent to `split` and `cat`, minus some abstractions that are internal to the filesystem. How the extents are laid on disk is also completely irrelevant once the model is loaded into ram (either the page cache if you're using mmap, or anon pages if you aren't).

**Reply by ngxson** [reply]
Thanks for pointing it out. Yeah I had a misconception about `mmap`. Let's consider my question as wrong question then.

**Comment by phymbert** [comment]
@ggerganov as we merged:
 - #6135
 
I am working on `llama_model_loader` to support split gguf in 14442d8315f5522d8d71cc20f55bdbda6f35bc72.

I need your advice to continue further on, should we have multiple `std::unique_ptr<llama_mmap> mapping` or  `llama_mmap` should support multiple input files ?
Thanks

**Reply by phymbert** [reply]
Well it looks less difficult to modify `llama_mmap` at least for posix. Keep you posted

**Reply by ngxson** [reply]
yeah for each file, I think you need to add a tuple (`llama_file`, `llama_mmap`, `gguf_context`, `ggml_context`).

Maybe you can create a new `struct llama_model_split` that hold these info. Then inside `llama_model_loader` you add a `std::vector<struct llama_model_split>`

**Reply by slaren** [reply]
I suspect using multiple instances of `llama_mmap` will be easier. For backends that support mmap like CPU and Metal, it will also need a different backend buffer for each file.

**Reply by phymbert** [reply]
Will a mmap with `MAP_ANONYMOUS` with all files mapped sequentially not work?
I think we need only one `ggml_context`, so I am also modifying `ggml` to support building a new `ggml_context *` from added tensors of split ggufs.

**Reply by phymbert** [reply]
@slaren right :+1: mapping multiple files to the same mapped memory requires pagesize aligment, as you told easier to have multiple `llama_map`. I finally managed to build the `ggml_context` without touching `ggml`. And manipuling mapped memory is too system dependent to ship this feature smoothly.

**Reply by phymbert** [reply]
@slaren As I understand `llama_mmap.unmap_fragment`  is always called with first=`gguf_get_data_offset(ctx_gguf)`, last=`first+ gguf_get_tensor_offset(ctx_gguf, last_tensors_idx) +  + ggml_nbytes(last_tensors)` - aside the case the `llama_progress_callback` fails. Could it be simplified by mapping only the tensor data in the gguf and unmapping all after tensors data are loaded to backend ? It will ease my implementation with multiple `llama_mmap`. Thanks.

**Reply by slaren** [reply]
The goal of `unmap_fragment` is to unmap the tensors that have been offloaded to VRAM and are not needed in system memory anymore. In practice it doesn't work very well, the input layer is never offloaded and the output layer is the last to be offloaded, and everything in between is kept mapped. So as it is, it only really works with full offload, as in that case the only mapped tensors are the input layer and everything else is unmapped. So it needs to be improved, but not removed completely.

**Reply by phymbert** [reply]
@slaren I implemented the multiple `llama_mmap` logic in 111be05fbba74255303c39bdd4e6e8cd18d8f26d. BUT I believe it cannot work because here tensors layers expect the memory to be contiguous:
https://github.com/ggerganov/llama.cpp/blob/91f8ad167dcd24b54615b468d9dd764ebe1d37ad/llama.cpp#L3166

So the same with multiple files fails with SIGV:
https://github.com/ggerganov/llama.cpp/blob/111be05fbba74255303c39bdd4e6e8cd18d8f26d/llama.cpp#L3314

@ggerganov your help is really welcomed here :), thanks

**Reply by phymbert** [reply]
Or I miss something on how `cur->data` is initialized ?

**Reply by slaren** [reply]
Tensors need to be contiguous in memory, but not layers.

**Reply by phymbert** [reply]
Hold on I broke something :/

**Reply by ngxson** [reply]
I've just have a look on the implementation. Seems like you keep all the current variables and just add some `std::map`. This may cause duplicated logic and at some case we don't know which variable is populated (and which is not).

I'd suggest that we move all the llama_file, llama_mmap,... to `std::set` or `std::vector`, then we process all `i_split` the same way (reduce the number of branch `n_split < 2`)

The sigsegv that you got maybe due to that: maybe some branchs access to `mapping` while it should be `split_mapping` for example.

**Reply by ngxson** [reply]
To me, only the part where we load metadata that we care which `i_split` it is. After that, we don't care from which split the tensor come from - so we should prevent checking `n_split < 2` in these cases.

**Reply by ngxson** [reply]
Another tips is maybe you can firstly make a non-mmap version that works (which should be easier), then bring back mmap after that.

**Reply by phymbert** [reply]
Thanks for your early review but how you can load tensor data to the backend if you dont know in which files/mapping/offset it is ? Metadata are not enough... Anyway I broke the branch where n_split < 2, so I am fixing this first.

**Reply by ngxson** [reply]
The offset of a tensor is already stored inside its `ctx_gguf`, so you can just store the `ctx_gguf` of the split into a std::map for example, which map name ==> ctx_gguf

Then we delegate the task of finding the offset to `file_offset`, as we already had this function. The modified version will looks like:

```cpp
    size_t file_offset(const char * name) const {
        const split_ctx_gguf = map_ctx_gguf[name]; // map name to context
        const int idx = gguf_find_tensor(split_ctx_gguf, name);

        if (idx < 0) {
            throw std::runtime_error(format("%s: tensor '%s' not found in the file", __func__, name));
        }

        return gguf_get_data_offset(ctx_gguf) + gguf_get_tensor_offset(ctx_gguf, idx);
    }
```

Doing this will reduce one occurrence where you need to duplicate the line `gguf_get_data_offset(ctx_gguf) + gguf_get_tensor_offset(ctx_gguf, idx)`

**Reply by phymbert** [reply]
I have a working version for both CUDA and CPU backend. Cleaning it up and opening a draft PR. Thanks all

**Comment by phymbert** [comment, marked as answer]
Completed in: 
- #6135
- #6187
- #6192

**Reply by ego** [reply]
Awesome job guys! 

Only one mention about arg `--split-max-size 3G` sometimes it produce largest files, but `--split-max-tensors` works like charm 👍

**Reply by ngxson** [reply]
@ego Can you let me know the model that you're using?

@phymbert I though about this bug before, maybe because I removed the code to calculate size of metadata because the code was quite too complicated. Maybe I'll need to fix this.

**Reply by phymbert** [reply]
We need tests <3

**Reply by ego** [reply]
Model is `CodeLlama-7b-Python` size ~26G

```bash
./gguf-split --split --split-max-size 3G ./models/CodeLlama-7b-Python/ggml-model-f32.gguf ./models/CodeLlama-7b-Python/splited/
```
and its produce only 3 files.

```bash
./main --version
version: 2584 (33a52448)
built with Apple clang version 15.0.0 (clang-1500.3.9.4) for arm64-apple-darwin23.4.0

./gguf-split --version
version: 2584 (33a52448)
built with Apple clang version 15.0.0 (clang-1500.3.9.4) for arm64-apple-darwin23.4.0
```

**Comment by trollkotze** [comment]
What's wrong with split and cat? How can anything be "smarter" than that?
This seems like an absolutely worthless complication that is now sadly becoming the new standard.

**Reply by ngxson** [reply]
Technical details: the new gguf-split tool splits the file at tensor level, not at byte level, so it never cuts a tensor in half (which will be the case when using split & cat). This allows us to:
- Load all shards without doing any `cat`, because a tensor never be cut in half
- Quantize a big model while keeping all the shards: https://github.com/ggerganov/llama.cpp/pull/6688
- ...

Also, this is the same approach used by safetensors.

Whether it's "smarter" or "complicated", it's up to you to judge. For me, it's just more useful this way.

**Reply by phymbert** [reply]
https://twitter.com/mishig25/status/1783108187119329301

**Reply by trollkotze** [reply]
Okay, I understand now. Sorry for being rude and obnoxious.
